### PR TITLE
Support complex objective functions

### PR DESCRIPTION
**Pull request by antoine-levitt**
See https://github.com/JuliaNLSolvers/Optim.jl/pull/440. Comments very welcome!

### CODE REVIEW

File: src/objective_types.jl [hunk 1]
```diff
@@ -17,31 +19,48 @@ type NonDifferentiable{T} <: AbstractObjective
...
+    NonDifferentiable(f, f(x_seed), copy(x_seed), [1], iscomplex)
+end
 
 # Used for objectives and solvers where the gradient is available/exists
 type OnceDifferentiable{T, Tgrad} <: AbstractObjective
     f
     g!
     fg!
-    f_x::T
+    f_x::Real
```
Comment (pkofod) [inline src/objective_types.jl]: This will mean that the type of f_x is never inferred

**Comment by pkofod** [comment]
I'm wondering if it isn't simply easier to dispatch on complex eltypes instead of checking some fields' equality to true? This is exactly the sort of thing you can avoid in Julia compared to say Matlab.

File: src/objective_types.jl (hunk 1 above)
Comment (antoine-levitt) [inline src/objective_types.jl]: Sorry, this is a typo, will fix.

**Comment by antoine-levitt** [comment]
Do you mean in functions such as OnceDifferentiable(f, g!, fg!, x_seed::AbstractArray), or more generally?

In the particular case of this function, there's less code duplication written in this way, and the performance is the same (I think) because the compile-time type of x_seed is known.

The main issue with type-stability in this approach is that the type of OnceDifferentiable is not enough to determine whether the optimization problem is real or complex. I haven't found any convenient way of doing this, though; we could have RealOnceDifferentiable and ComplexOnceDifferentiable, both subtypes of OnceDifferentiable, but this gets pretty messy (have to do this for Non-, Once- and Twice-) and duplicates the structure. Possibly this is a good use for traits? As I said, I'm a beginner at this, so advice appreciated.

File: src/objective_types.jl [hunk 2]
```diff
@@ -12,17 +14,26 @@ function fix_order(storage_input, x_input, fun!, fun!_msg)
     end
 end
 # Used for objectives and solvers where no gradient is available/exists
-type NonDifferentiable{T} <: AbstractObjective
+type NonDifferentiable{T<:Real} <: AbstractObjective
```
Comment (pkofod) [inline src/objective_types.jl]: What does this buy us? I'm sure there's a reason, as you did it below as well.

File: src/objective_types.jl [hunk 3]
```diff
@@ -12,17 +14,26 @@ function fix_order(storage_input, x_input, fun!, fun!_msg)
...
+    iscomplex = eltype(x_seed) <: Complex
+    if iscomplex
+        x_seed = complex_to_real(x_seed)
+    end
+    NonDifferentiable(f, f(x_seed), copy(x_seed), [1], iscomplex)
+end
 
 # Used for objectives and solvers where the gradient is available/exists
-type OnceDifferentiable{T, Tgrad} <: AbstractObjective
+type OnceDifferentiable{T<:Real, Tgrad} <: AbstractObjective
```
Comment (pkofod) [inline src/objective_types.jl]: same as above

**Comment by pkofod** [comment]
Coming back to this...

I'm not really sure what we're trying to do here. I mean, what is the reason why we need to go back and forth all the time? I may be wrong, but I cannot see how this is a no-op for the normal case?

Don't get me wrong, I can see why supporting complex inputs can be helpful, but you're the first person mentioning it, so I'd prefer to keep the added complexity as low as possible.

What goes wrong by simply keeping `x` a Complex for example?

File: src/objective_types.jl (hunk 2 above)
Comment (antoine-levitt) [inline src/objective_types.jl]: Nothing, it's just a type assertion because it does not make sense for T to be anything else

**Comment by antoine-levitt** [comment]
See comment in the issue. It's possible to keep x complex, but
- it's not compatible with second order solvers 
- the code needs to be modified in an error prone way (eg if you forget a real() it will not crash but will degrade convergence) 
- I think bfgs has to convert to 2n anyway (not sure) 

This implementation is a bit verbose and it would be nice to do without, but it's pretty localized and less likely to introduce bugs. 

It's not a noop because of type inference issues only (I think it's possible to make it type inferrable, but not sure what's the best way). In particular, complex to real and back does not make a copy.

**Comment by pkofod** [comment]
```
real_to_complex(x::AbstractArray) = x
real_to_complex{T<:Complex}(x::AbstractArray{T} = ...
```
?

**Comment by antoine-levitt** [comment]
The second one is OK, the first one not (real representation of complex arrays are 2 x original_dims, with no distinguishing characteristic than that)

**Comment by pkofod** [comment]
> The second one is OK, the first one not (real representation of complex arrays are 2 x original_dims, with no distinguishing characteristic than that)

sorry, I deleted my comment just before you answered :)

File: src/objective_types.jl (hunk 2 above)
Comment (pkofod) [inline src/objective_types.jl]: Some types might have all the necessary methods but might not be `<:Real`, and so cannot be paased

Comment (antoine-levitt) [inline src/objective_types.jl]: Really? Aren't the necessary operations (arithmetic ones) precisely what being a `Real` means? Any type that implements these should be a subtype of Real.

**Comment by pkofod** [comment]
Okay, I get it now. I think we need only handle this in Optim. I have very little time these days, but the design I have in mind is something like this:

We do like you propose: convert the seed to an array with double the number of elements for internal use and convert back to complex when evaluations happen. We use the same functions you propose here, but, we can simply make a constructor in Optim that handles complex cases, and creates a new function on the fly

```
f_comp(x) = f(real_to_complex(x))
```

I find this better as we don't  add another field to a type (I think we have many as it is...), we don't touch NLSolversBase, and it doesn't touch the normal case at all (that is, no code changes and no contextually "weird" `real_to_complex` names in code that could confuse people).

**Comment by pkofod** [comment]
I mean... We could potentially also do it here, but does it make sense?

**Comment by antoine-levitt** [comment]
Ok, but there are several places where you want to know whether you are dealing with a complex array (manifold projection, return value, possibly preconditioning...) so the information that it is complex has to be stored somewhere anyway...

**Comment by pkofod** [comment]
> Ok, but there are several places where you want to know whether you are dealing with a complex array (manifold projection, return value, possibly preconditioning...) so the information that it is complex has to be stored somewhere anyway...

Alright, I guess I'll have to live with it :) I fixed your conflicting change, so you may want to pull the latest commit before working on this. I can't seem to find a `complex_to_real` fallback ?

**Comment by antoine-levitt** [comment]
I'm not sure, what do you mean by that?

**Comment by pkofod** [comment]
> What do you mean by complex_to_real fallback?

Nothing, I realized afterwards that this was not used like I thought it was!

**Comment by pkofod** [comment]
If you're using the `where` syntax, you need to drop v0.5 support in REQUIRE and update the travis and appveyor files accordingly.

**Comment by antoine-levitt** [comment]
Done (also removed some obsolete stuff)

**Comment by pkofod** [comment]
Gotta say, I still don't understand why `iscomplex` is needed... Just dispatch on `AbstractArray{T}`  with `T<:Complex` and a fallback? Can't see a single place where `iscomplex` is used where it is actually necessary (here or Optim PR), but it's very much possible that I'm just slow.

**Comment by antoine-levitt** [comment]
When you're in a solver and you want to call f (or a retraction on the manifold, or something), how do you know whether you should convert x (which is a real 2 x something array) to a complex array or not?

**Comment by antoine-levitt** [comment]
I added some trickery so that real_to_complex gets inferred. With this, real_to_complex() is really a no-op in the real case (I checked that using `@code_warntype`). In the complex case it's still not fully inferrable because the number of dimensions is not inferred, if somebody knows how to do it that'd be great!

**Comment by pkofod** [comment]
Thanks!, I'll cook up a new version of NLSolversBase and then I'll merge the Optim PR

**Pull request merged**